FEA Introduce `PairwiseDistances`

jjerphan opened the pull request:
#### Reference Issues/PRs

Relates to https://github.com/scikit-learn/scikit-learn/issues/22587.

#### What does this implement/fix? Explain your changes.

This adds a new back-end to `pairwise_distances` computations using `PairwiseDistances` without any reduction.


#### Any other comments?

TODO:
 - [x] merge https://github.com/scikit-learn/scikit-learn/pull/22666/
 - [x] agree on a resolution for https://github.com/scikit-learn/scikit-learn/issues/24502 and implements its declination for this back-end
 - [x] perform benchmarks

jjerphan committed:
Introduce PairwiseDistances

jjerphan committed:
Merge branch 'main' into maint/pairwise_distances-pdr-backend

jjerphan committed:
WIP

jjerphan committed:
Merge branch 'main' into maint/pairwise_distances-pdr-backend

jjerphan committed:
fixup! Introduce PairwiseDistances

jjerphan committed:
Post-merge fix

ಠ_ರೃ

jjerphan committed:
Merge branch 'main' into feat/pairwise_distances-pdr-backend

jjerphan committed:
Merge branch 'main' into feat/pairwise_distances-pdr-backend

jjerphan commented:
Failures on this PR might are fixed by https://github.com/scikit-learn/scikit-learn/pull/23990.

jjerphan committed:
Merge branch 'main' into feat/pairwise_distances-pdr-backend

jjerphan committed:
Merge branch 'main' into feat/pairwise_distances-pdr-backend

jjerphan committed:
fixup! Merge branch 'main' into feat/pairwise_distances-pdr-backend

Switch condition for safety

jjerphan committed:
Merge branch 'main' into feat/pairwise_distances-pdr-backend

jjerphan committed:
Do not offset by X_start and Y_start

jjerphan committed:
Do not progated metric_kwargs unneedlessly

jjerphan committed:
Merge branch 'main' into feat/pairwise_distances-pdr-backend

jjerphan committed:
Use the proper vectors' indices

jjerphan committed:
Port PairwiseDistances to support float32 datasets

jjerphan committed:
Simplify indices

jjerphan committed:
Adapt implementations for a previous 'sqeuclidean' specification

jjerphan committed:
TST Downcast distance matrix to float32

jjerphan committed:
Adapt instanciation

jjerphan committed:
Use PairwiseDistances as a back-end for haversine_distances

jjerphan committed:
Use PairwiseDistances as a back-end for manhattan_distances

jjerphan committed:
Use PairwiseDistances as a back-end for euclidean_distances

Comes with minor adaptations

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/setup.py:
```diff
@@ -31,7 +33,9 @@ def configuration(parent_package="", top_path=None):
         "_datasets_pair.pyx",
         "_gemm_term_computer.pyx",
         "_base.pyx",
+        "_pairwise_distances.pyx",
```
```suggestion
```

jjerphan committed:
Remove duplicated line

jjerphan commented on sklearn/metrics/pairwise.py:
```diff
@@ -1663,32 +1673,6 @@ def _check_chunk_size(reduced, chunk_size):
         )
 
 
-def _precompute_metric_params(X, Y, metric=None, **kwds):
-    """Precompute data-derived metric parameters if not provided."""
-    if metric == "seuclidean" and "V" not in kwds:
-        # There is a bug in scipy < 1.5 that will cause a crash if
-        # X.dtype != np.double (float64). See PR #15730
-        dtype = np.float64 if sp_version < parse_version("1.5") else None
-        if X is Y:
-            V = np.var(X, axis=0, ddof=1, dtype=dtype)
-        else:
-            raise ValueError(
-                "The 'V' parameter is required for the seuclidean metric "
-                "when Y is passed."
-            )
-        return {"V": V}
-    if metric == "mahalanobis" and "VI" not in kwds:
-        if X is Y:
-            VI = np.linalg.inv(np.cov(X.T)).T
-        else:
-            raise ValueError(
-                "The 'VI' parameter is required for the mahalanobis metric "
-                "when Y is passed."
-            )
-        return {"VI": VI}
-    return {}
-
-
```
This has been moved in `sklearn/metrics/_pairwise_distances_reduction/_pairwise_distances.pyx.tp` to avoid cyclic references.

jjerphan commented on sklearn/neighbors/_nca.py:
```diff
@@ -486,7 +486,7 @@ def _loss_grad_lbfgs(self, transformation, X, same_class_mask, sign=1.0):
 
         # Compute softmax distances
-        p_ij = pairwise_distances(X_embedded, squared=True)
+        p_ij = pairwise_distances(X_embedded, metric="sqeuclidean")
```
This is equivalent, but preferable IMO.

ogrisel replied:
I agree this is more explicit.

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_pairwise_distances.pyx.tp:
```diff
@@ -0,0 +1,373 @@
+        )
+
+    def _finalize_results(self):
+        # If X is Y, then catastrophic cancellation might
+        # have occurred for computations of term on the diagonal
+        # which must be null. We enforce nullity of those term
+        # by zeroing the diagonal.
+        distance_matrix = np.asarray(self.pairwise_distances_matrix)
+        if self.datasets_pair.X_is_Y:
+            np.fill_diagonal(distance_matrix, 0.)
```
Potentially, we might want (remove this and) add a guard directly after computing distances to prevent negative zeros and NaN [as done in the Euclidean Variant](https://github.com/jjerphan/scikit-learn/blob/05b5ae7774425c4e07004fd0967a85c71c124d4e/sklearn/metrics/_pairwise_distances_reduction/_pairwise_distances.pyx.tp#L357-L361).

If removing this part, then `DatasetsPair.X_is_Y` must also be removed.

What do you think?

jeremiedbb replied:
`_euclidean_distances` does prevent negative values. I think we should keep that behavior

jeremiedbb replied:
We should generalize that to all metrics (if they are true metrics). Since this not used only in `pairwise_distances` I guess we have to do that here to avoid code duplication

jjerphan replied:
What I meant is that we should also propagate the [guard against negative zeros and NaNs in `EuclideanPairwiseDistances`](https://github.com/jjerphan/scikit-learn/blob/05b5ae7774425c4e07004fd0967a85c71c124d4e/sklearn/metrics/_pairwise_distances_reduction/_pairwise_distances.pyx.tp#L357-L361) to other `Euclidean` variant because they are not  all protected against NaNs and negative zeros.

jjerphan commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -164,7 +164,7 @@ def test_binary_search_neighbors():
     random_state = check_random_state(0)
     data = random_state.randn(n_samples, 2).astype(np.float32, copy=False)
-    distances = pairwise_distances(data)
+    distances = pairwise_distances(data).astype(np.float32, copy=False)
```
This change is introduced for `_binary_search_perplexity` to run (it does not accept float64 array).

Must `pairwise_distances` preserve `dtype`? If so, should we downcast in this case?

jeremiedbb replied:
> Must pairwise_distances preserve dtype

That's not a simple question :(
Currently it does sometimes but not always. I'm opening an issue about that

jeremiedbb replied:
https://github.com/scikit-learn/scikit-learn/issues/24502

jjerphan replied:
Based on discussion in https://github.com/scikit-learn/scikit-learn/issues/24502,  cfc145ab51fcfd7203fd22f8ede08529331c1125 makes `PairwiseDistances` preserves dtypes, hence reverting this now-useless change.

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_pairwise_distances.pyx.tp:
```diff
@@ -0,0 +1,373 @@
+            chunk_size=self.chunk_size,
+        )
+
+        if metric_kwargs is not None and metric_kwargs.get("Y_norm_squared", None) is not None:
+            self.Y_norm_squared = metric_kwargs.pop("Y_norm_squared")
+        else:
+            self.Y_norm_squared = _sqeuclidean_row_norms{{name_suffix}}(
+                datasets_pair.Y,
+                self.effective_n_threads,
+            )
+
+        if metric_kwargs is not None and metric_kwargs.get("X_norm_squared", None) is not None:
+            self.X_norm_squared = metric_kwargs.pop("X_norm_squared")
+        else:
+            # Do not recompute norms if datasets are identical.
+            self.X_norm_squared = (
+                self.Y_norm_squared if self.datasets_pair.X_is_Y else
+                _sqeuclidean_row_norms{{name_suffix}}(
+                    datasets_pair.X,
+                    self.effective_n_threads,
+                )
+            )
```
Note: this is safer than what the equivalent lines in other `Euclidean*` specialisation.

jeremiedbb replied:
Can you explain more, I did not understand this note

jjerphan replied:
This unpacking in `EuclideanPairwiseDistances` for `{X,Y}_norm_squared` is safer than the one in `EuclideanArgKmin` and `EuclideanRadiusNeighbors`: this one checks for `None` better, where the others do not.

jjerphan replied:
> So you mean we should update the order reducing variants to follow the same strategy in a follow-up PR?

Yes, so as to prevent issues like the current ones on the CI as you mentioned in https://github.com/scikit-learn/scikit-learn/pull/23958#issuecomment-1260531381

jjerphan commented on setup.cfg:
```diff
@@ -79,6 +79,8 @@ ignore =
     sklearn/metrics/_pairwise_distances_reduction/_datasets_pair.pyx
     sklearn/metrics/_pairwise_distances_reduction/_gemm_term_computer.pxd
     sklearn/metrics/_pairwise_distances_reduction/_gemm_term_computer.pyx
+    sklearn/metrics/_pairwise_distances_reduction/_pairwise_distances.pxd
+    sklearn/metrics/_pairwise_distances_reduction/_pairwise_distances.pyx
```
Side-note: it would be nice not to have duplication with identical lines in `.gitignore`, yet I am not sure.

jjerphan reviewed (commented):
A few notes for review discussions.

jjerphan committed:
Merge branch 'scikit-learn:main' into feat/pairwise_distances-pdr-backend

jjerphan committed:
Merge branch 'main' into feat/pairwise_distances-pdr-backend

Also adapt the logic and document the semantic around
`DatasetsPair.X_is_Y`.

jjerphan committed:
TST Remove checks on errors now that minkowski with sparse data is supported

jjerphan committed:
TST PairwiseDistances factory methods

jeremiedbb commented on sklearn/metrics/pairwise.py:
```diff
@@ -1982,6 +1966,17 @@ def pairwise_distances(
             % (metric, _VALID_METRICS)
         )
 
+    if PairwiseDistances.is_usable_for(X, X if Y is None else Y, metric=metric):
+        # This is an adaptor for one "sqeuclidean" specification.
+        # For this backend, we can directly use "sqeuclidean".
+        if kwds.get("squared", False) and metric == "euclidean":
+            metric = "sqeuclidean"
+            kwds = {}
+
+        return PairwiseDistances.compute(
+            X, X if Y is None else Y, metric=metric, metric_kwargs=kwds
+        )
```
And I wonder if we can now remove some parts below that are now unreachable ?

ogrisel replied:
code coverage should tell us if any Python part have become unreachable, right? It's weird that is does not show up in the CI status report (maybe because of failing tests?) but here is the report and it seems that we have a bit of dead code:

https://app.codecov.io/gh/scikit-learn/scikit-learn/pull/23958

jjerphan replied:
I think parts of the previous implementations will still be used in some context and still will be reachable/covered. For instance boolean metrics aren't yet supported.

jjerphan replied:
Moreover, if either `X` or `Y` is F-contiguous, then those implementations aren't usable. 

I think we still need to maintain the old implementations because forcing F-contiguous arrays to be C-contiguous might be costly. What do you think?

jjerphan replied:
Further more when inputs are lists of lists of number, we currently need to use the parts bellow because do not get converted to numpy arrays.

See this RFC: https://github.com/scikit-learn/scikit-learn/issues/24745

jeremiedbb reviewed (commented):
Implementation details looks good. Here are a few comments more focus on integration in the `pairwise` module

jjerphan committed:
Merge branch 'main' into feat/pairwise_distances-pdr-backend

jjerphan committed:
MAINT Remove _sparse_manhattan

Co-authored-by: Jérémie du Boisberranger <jeremiedbb@users.noreply.github.com>

jjerphan committed:
DOC Remove comment regarding X_is_Y

Co-authored-by: Jérémie du Boisberranger <jeremiedbb@users.noreply.github.com>

jjerphan committed:
Preserve dtype for PairwiseDistances

Following discussions in:
https://github.com/scikit-learn/scikit-learn/issues/24502

jjerphan commented on sklearn/metrics/pairwise.py:
```diff
@@ -933,6 +940,9 @@ def manhattan_distances(X, Y=None, *, sum_over_features=True):
     """
     X, Y = check_pairwise_arrays(X, Y)
 
+    if sum_over_features and PairwiseDistances.is_usable_for(X, Y, metric="manhattan"):
+        return PairwiseDistances.compute(X, Y, metric="manhattan")
```
`_sparse_manhattan` is still needed because it allow supporting sparse matrices that aren't CSR by converting them to this format. Such conversation is currently not done for case for `PairwiseDistancesReduction` which only accepts a certain kind of CSR matrices:

https://github.com/scikit-learn/scikit-learn/blob/53d6c07f072b5cfdc4f1e687243f916555dbf174/sklearn/metrics/_pairwise_distances_reduction/_dispatcher.py#L105-L141

I think the removal of `_sparse_manhattan` has to be postponed because as might we first need another PR broaden the kind of supported sparse matrices for `PairwiseDistancesReduction`.

jeremiedbb replied:
But it's not ``_sparse_manhattan`` that does the conversion. In ``manhattan_distances`` we check if one is sparse and if so, we convert both to CSR, then call ``_sparse_manhattan``. So what I meant is that we can't make this check/conversion first and then call PairwiseDistancesReduction instead of ``_sparse_manhattan``

jjerphan replied:
Ah yes, 259edc1d64660262ebfa54101d64a474fc52666d treats your remarks.

jjerphan commented on sklearn/metrics/pairwise.py:
```diff
@@ -1982,6 +1966,17 @@ def pairwise_distances(
             % (metric, _VALID_METRICS)
         )
 
+    if PairwiseDistances.is_usable_for(X, X if Y is None else Y, metric=metric):
+        # This is an adaptor for one "sqeuclidean" specification.
+        # For this backend, we can directly use "sqeuclidean".
+        if kwds.get("squared", False) and metric == "euclidean":
+            metric = "sqeuclidean"
+            kwds = {}
+
+        return PairwiseDistances.compute(
+            X, X if Y is None else Y, metric=metric, metric_kwargs=kwds
+        )
```
That's right. We forgot to document the change of some interfaces' behavior regarding `n_jobs` when they are relying on `PairwiseDistancesReductions`. That's itself an item of #22587.

jjerphan replied:
I am planning to do a PR including docstrings' updates and a documentation regarding `n_jobs` to inform/educate users w.r.t this change. I guess such a PR is more urgent and important than this one for 1.2 as there's currently a discrepancy between what is documented for computations parallelism and how computations effectively execute.

jjerphan committed:
Merge branch 'main' into feat/pairwise_distances-pdr-backend

jjerphan committed:
TST Remove TODO now that dtypes are preserved

jjerphan committed:
DOC Remove and adapt some comments

jjerphan committed:
Use PairwiseDistance as a back-end for _euclidean_distances

jjerphan committed:
MAINT Keep classmethods at the top

ogrisel commented:
CI is complaining about `ValueError: Buffer has wrong number of dimensions (expected 1, got 2)` in `test_euclidean_distances_with_norms[float64-dense]`.

There is also a `rtol` to adjust in `test_euclidean_distances_extreme_values`.

ogrisel commented on sklearn/metrics/pairwise.py:
```diff
@@ -933,21 +948,22 @@ def manhattan_distances(X, Y=None, *, sum_over_features=True):
             raise TypeError(
                 "sum_over_features=%r not supported for sparse matrices"
                 % sum_over_features
             )
```
Code coverage says that is also not reachable:

https://app.codecov.io/gh/scikit-learn/scikit-learn/pull/23958

but maybe it's because it's not up to date because of the remaining failing tests?

Not sure...

jjerphan replied:
`manhattan_distances` with `sum_over_features=False` does not return distances, looks unused, and is also unlikely used by users.

This is out of the scope of this PR, but to me the behavior for `sum_over_features=True` should be the unique one for `manhattan_distances`; in this case, `sum_over_features` would be removed this from the public API after a deprecation cycle. WDYT?

ogrisel replied:
> in this case, sum_over_features would be removed this from the public API after a deprecation cycle. WDYT?

If it is actually never used in scikit-learn, +1 for deprecation.

jjerphan replied:
I opened https://github.com/scikit-learn/scikit-learn/issues/24597 accordingly.

ogrisel reviewed (commented):
Here is a pass of review, mostly nits. Assuming the benchmarks do not show regression and tests can be fixed with a good coverage, this looks good to me.

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_pairwise_distances.pyx.tp:
```diff
@@ -0,0 +1,385 @@
+    """EuclideanDistance-specialized {{name_suffix}}bit implementation for PairwiseDistances."""
+
+    @classmethod
+    def is_usable_for(cls, X, Y, metric) -> bool:
```
IIRC  I typed `is_usable_for` a long time ago since it was Cython, but it was then moved to Python and can thus can be removed. I am +0 as well.

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_pairwise_distances.pyx.tp:
```diff
@@ -0,0 +1,385 @@
+                f"usable for this case (EuclideanPairwiseDistances{{name_suffix}}) and will be ignored.",
+                UserWarning,
+                stacklevel=3,
+            )
```
+1. I can do this as part of another smaller PR.

ogrisel replied:
I would rather avoid introducing a new warning that we want to actually get rid off, so better do this as part of the current PR.

jjerphan replied:
My remark was not clear and I wanted to say that I can remove the warning for this implementation and remove existing warnings as part of another PR such as #24542.

jjerphan committed:
Safely pack {X,Y}_squared_norms

jjerphan committed:
TST Adapt test_euclidean_distances_extreme_values

This:
 - decreases the number of features by an order to magnetude
   because in the case of float32, the vectors gets entirely
   copied for the upcast to float64. This might use too much
   memory and crash the program
 - this now accepts the previously xfail parametrisation case
   by setting on absolute error (seen we are comparing small
   values)

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_dispatcher.py:
```diff
@@ -165,6 +169,127 @@ def compute(
+            )
+
+        raise ValueError(
+            "Only float64 or float32 datasets pairs are supported at this time, "
+            f"got: X.dtype={X.dtype} and Y.dtype={Y.dtype}."
```
I agree, but do you instead mean:
```suggestion
            "But datasets should be provided with identical data types, "
            f"but currently: X.dtype={X.dtype} and Y.dtype={Y.dtype}."
```

?

jjerphan replied:
I agree. Would this simpler change work for you?
```suggestion
            "Only float64 or float32 datasets pairs are supported, "
            f"got: X.dtype={X.dtype} and Y.dtype={Y.dtype}."
```

jjerphan commented on sklearn/metrics/pairwise.py:
```diff
@@ -933,21 +948,22 @@ def manhattan_distances(X, Y=None, *, sum_over_features=True):
-        return D
-
     if sum_over_features:
         return distance.cdist(X, Y, "cityblock")
```
If X and Y are lists of lists of numbers (which is do support yet something do not convert to numpy arrays), then this code will be called.

More generally, sometimes lists of lists of numbers are passed down to Cython implementations even after inputs being validated. We have not yet decided how to treat those cases.

ogrisel replied:
Thanks for the clarification.

Can't we just make sure to convert to 2D numpy arrays in this the input validation step of the public methods / functions instead?

If not or it's too far reaching to be considered being done as part of this PR maybe it would be worth adding a TODO inline comment to state that this branch is still needed for this specific case and we might want to turn those list-of-lists to numpy array arrays in the future.

jjerphan replied:
I think it's worth changing the logic regarding list-of-lists in another PR because this change would impact a lot of implementations.

In the meantime, I will just add a inline TODO comment as you suggested.

jjerphan committed:
DOC Improve docstrings

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
Simplify manhattan_distances

jjerphan committed:
fixup! DOC Improve docstrings

jjerphan committed:
fixup! TST Adapt test_euclidean_distances_extreme_values

ogrisel commented on sklearn/metrics/pairwise.py:
```diff
@@ -337,6 +340,18 @@ def _euclidean_distances(X, Y, X_norm_squared=None, Y_norm_squared=None, squared
     """
+    metric = "sqeuclidean" if squared else "euclidean"
+    if PairwiseDistances.is_usable_for(X, Y, metric):
+        metric_kwargs = {
+            "X_norm_squared": np.ravel(X_norm_squared)
+            if X_norm_squared is not None
+            else X_norm_squared,
+            "Y_norm_squared": np.ravel(Y_norm_squared)
+            if Y_norm_squared is not None
+            else Y_norm_squared,
+        }
```
nit:

```suggestion
        metric_kwargs = {
            "X_norm_squared": np.ravel(X_norm_squared)
            if X_norm_squared is not None else None,
            "Y_norm_squared": np.ravel(Y_norm_squared)
            if Y_norm_squared is not None else None,
        }
```

ogrisel reviewed (commented):
Some more feedback below:

jjerphan committed:
Apply review comments

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
Merge branch 'main' into feat/pairwise_distances-pdr-backend

jjerphan committed:
Rework poping {X,Y}_norm_squared in EuclideanPairwiseDistances

jjerphan committed:
Rework poping {X,Y}_norm_squared in DatasetsPair.get_for

jjerphan commented:
Using `perf(1)`, we can see that when running `pairwise_distances(X, Y, dist="manhattan")`, 98% of the time is spent in computing distances, i.e. in `ManhattanDistance{32,64}.dist`:

https://github.com/scikit-learn/scikit-learn/blob/e41753ebd57c44ae91b389f190c43ddc0b384a75/sklearn/metrics/_dist_metrics.pyx.tp#L1110-L1120

Looking at cycles spent on the assembly code line, we can see that loop unrolling + SIMD is used:


```diff
Percent│
       │    Disassembly of section .text:
       │
       │    00000000000104b0 <__pyx_f_7sklearn_7metrics_13_dist_metrics_17ManhattanDistance_dist>:
       │    __pyx_f_7sklearn_7metrics_13_dist_metrics_17ManhattanDistance_dist():
  0.30 │      test   %rcx,%rcx
  0.04 │    ↓ jle    40
  0.00 │      movq   __pyx_k_C+0x6b4b,%xmm2
  0.01 │      xor    %eax,%eax
       │      pxor   %xmm1,%xmm1
  0.00 │      nop
  0.00 │18:   movsd  (%rsi,%rax,8),%xmm0
  0.28 │      subsd  (%rdx,%rax,8),%xmm0
  0.25 │      add    $0x1,%rax
  0.04 │      andpd  %xmm2,%xmm0
  0.18 │      addsd  %xmm0,%xmm1
+98.76 │      cmp    %rax,%rcx
  0.10 │    ↑ jne    18
  0.00 │      movapd %xmm1,%xmm0
  0.04 │    ← retq
       │      nop
       │40:   pxor   %xmm1,%xmm1
       │      movapd %xmm1,%xmm0
       │    ← retq   
```

It's insightful to see how it's done, and to assess that 98.76% the time here is spent checking if `j` has reached the range bound, `size`.

ogrisel commented:
As discussed in real life, it might be interesting to see of the chunking is not detrimental for this operation: indeed chunking will cause non-contiguous writing of the distance values in the distance matrix output array.

It might be worthwhile to conduct dedicated benchmarks to use a chunk size of `1` to see if contiguous writing is beneficial or not.

jjerphan commented on sklearn/metrics/pairwise.py:
```diff
@@ -337,6 +340,20 @@ def _euclidean_distances(X, Y, X_norm_squared=None, Y_norm_squared=None, squared
+
+    # XXX: the following code is still used for list-of-lists of numbers which
+    # aren't converted to numpy arrays in validation steps done in `check_array`.
+    # TODO: convert list-of-lists to numpy arrays in `check_array`.
```
```suggestion
    # TODO: convert list-of-lists to numpy arrays in `check_array`.
    # See: https://github.com/scikit-learn/scikit-learn/issues/24745
```

jjerphan commented on sklearn/metrics/pairwise.py:
```diff
@@ -868,6 +885,14 @@ def haversine_distances(X, Y=None):
+
+    # XXX: the following code is still used for list-of-lists of numbers which
+    # aren't converted to numpy arrays in validation steps done in `check_array`.
+    # TODO: convert list-of-lists to numpy arrays in `check_array`.
+
```
```suggestion
    # TODO: convert list-of-lists to numpy arrays in `check_array`.
    # See: https://github.com/scikit-learn/scikit-learn/issues/24745

```

jjerphan commented on sklearn/metrics/pairwise.py:
```diff
@@ -933,20 +958,29 @@ def manhattan_distances(X, Y=None, *, sum_over_features=True):
+    # XXX: the following code is still used for list-of-lists of numbers which
+    # aren't converted to numpy arrays in validation steps done in `check_array`
+    # and for supporting `sum_over_features` which we should probably remove.
+    # TODO: convert list-of-lists to numpy arrays in `check_array`.
```
```suggestion
    # TODO: convert list-of-lists to numpy arrays in `check_array`.
    # See: https://github.com/scikit-learn/scikit-learn/issues/24745
```

jjerphan commented on sklearn/metrics/pairwise.py:
```diff
@@ -2001,6 +2020,10 @@ def pairwise_distances(
     else:
+        # XXX: the following code is still used for list-of-lists of numbers which
+        # aren't converted to numpy arrays in validation steps done in `check_array`.
+        # TODO: convert list-of-lists to numpy arrays in `check_array`.
```
```suggestion
        # TODO: convert list-of-lists to numpy arrays in `check_array`.
        # See: https://github.com/scikit-learn/scikit-learn/issues/24745
```

jjerphan reviewed (commented):
I just propose to add some references to https://github.com/scikit-learn/scikit-learn/issues/24745 in TODO comments.

jjerphan committed:
DOC Add references to #24745

jjerphan committed:
Merge branch 'main' into feat/pairwise_distances-pdr-backend

jjerphan commented:
> It might be worthwhile to conduct dedicated benchmarks to use a chunk size of 1 to see if contiguous writing is beneficial or not.

My first intuition is that simply setting `chunk_size=1` might help but we will still relies on a rather to complicated scheduling for this case: some details like the heuristic will be adapted, we might spend a lot of time jumping in many places due to singletonic chunks, etc..

I think `_parallel_on_{X,Y}` in this case can be made relatively simpler (namely it would just generalised [the current `_sparse_manhattan`](https://github.com/scikit-learn/scikit-learn/blob/08275348cd2c186386d848bbe0099478049a4589/sklearn/metrics/_pairwise_fast.pyx#L38-L105) logic).

What do you think?

jjerphan committed:
Merge remote-tracking branch 'upstream/main' into feat/pairwise_distances-pdr-backend

jjerphan commented:
I am removing it from the 1.2 milestone as it needs more thought and thread-offs' assessments.

jjerphan commented:
Turning this into a draft for two reasons; to me:
 - this brings less value to end-users compared to other work (some of which is part of  #22587) that has been started in the meantime
 - there exists alternative implementations that are more lightweight and that do not necessitate the whole infrastructure for the distance matrix chunking; see https://github.com/scikit-learn/scikit-learn/pull/23958#issuecomment-1294676589

jjerphan commented:
Closing since this has been superseded by #25561.
